src/core/ddsc/src/dds_reader.c

```diff
@@ -497,70 +571,46 @@ static dds_entity_t dds_create_reader_int (dds_entity_t participant_or_subscribe
-     are removed before removing the DDSI participant. So at this point, within
-     the subscriber lock, we can assert that the participant exists. */
-  assert (pp != NULL);
-
-#ifdef DDS_HAS_SECURITY
-  /* Check if DDS Security is enabled */
-  if (q_omg_participant_is_secure (pp))
-  {
-    /* ask to access control security plugin for create reader permissions */
-    if (!q_omg_security_check_create_reader (pp, gv->config.domainId, tp->m_name, rqos))
-    {
-      rc = DDS_RETCODE_NOT_ALLOWED_BY_SECURITY;
-      goto err_not_allowed;
-    }
-  }
-#endif
+  /* get the autoenable setting of the parent
+   * this setting determines if the entity must be enabled or not */
+  dds_qget_entity_factory(sub->m_entity.m_qos, &autoenable);
 
   /* Create reader and associated read cache (if not provided by caller) */
   struct dds_reader * const rd = dds_alloc (sizeof (*rd));
   const dds_entity_t reader = dds_entity_init (&rd->m_entity, &sub->m_entity, DDS_KIND_READER, false, rqos, listener, DDS_READER_STATUS_MASK);
-  rd->m_sample_rejected_status.last_reason = DDS_NOT_REJECTED;
   rd->m_topic = tp;
   rd->m_wrapped_sertopic = (tp->m_stype->wrapped_sertopic != NULL) ? 1 : 0;
-  rd->m_rhc = rhc ? rhc : dds_rhc_default_new (rd, tp->m_stype);
-  if (dds_rhc_associate (rd->m_rhc, rd, tp->m_stype, rd->m_entity.m_domain->gv.m_tkmap) < 0)
-  {
-    /* FIXME: see also create_querycond, need to be able to undo entity_init */
-    abort ();
-  }
+  rd->m_rhc = rhc;
   dds_entity_add_ref_locked (&tp->m_entity);
-
   /* FIXME: listeners can come too soon ... should set mask based on listeners
      then atomically set the listeners, save the mask to a pending set and clear
      it; and then invoke those listeners that are in the pending set */
+  rd->m_entity.m_iid = ddsi_iid_gen();
   dds_entity_init_complete (&rd->m_entity);
-
-  rc = new_reader (&rd->m_rd, &rd->m_entity.m_guid, NULL, pp, tp->m_name, tp->m_stype, rqos, &rd->m_rhc->common.rhc, dds_reader_status_cb, rd);
-  assert (rc == DDS_RETCODE_OK); /* FIXME: can be out-of-resources at the very least */
-  thread_state_asleep (lookup_thread_state ());
-
-  rd->m_entity.m_iid = get_entity_instance_id (&rd->m_entity.m_domain->gv, &rd->m_entity.m_guid);
-  dds_entity_register_child (&sub->m_entity, &rd->m_entity);
-
+  if (autoenable) {
+    rc = dds_reader_enable (&rd->m_entity);
+  }
+  dds_entity_register_child (rd->m_entity.m_parent, &rd->m_entity);
   dds_topic_allow_set_qos (tp);
   dds_topic_unpin (tp);
   dds_subscriber_unlock (sub);
+  /* Destroy the reader and return DDS_RETCODE_NOT_ALLOWED_BY_SECURITY
+   * when no credentials could be obtained in case of auto_enabled=true */
+  if (!dds_entity_creation_allowed (rc))
+    goto err_not_allowed;
   return reader;
 
-#ifdef DDS_HAS_SECURITY
 err_not_allowed:
-  thread_state_asleep (lookup_thread_state ());
-#endif
+  (void) dds_delete (reader);
+  goto del_implicit_sub;
```

Comment:
I'm not against `goto` for error handling, but this moves just a little too far in the direction of "unstructured" programming to my taste!

(Also for the writer.)